staking/staking_hub.go

```diff
@@ -0,0 +1,419 @@
+func (sh *StakingHub) ResetStake(address types.Address) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) {
+		delete(sh.StakingMap, address)
+	}
+
+	sh.log(
+		fmt.Sprintf("Stake reset:\t%s", address.String()),
+		logInfo,
+	)
+}
+
+// GetStakedBalance returns an accounts staked balance if it is a staker.
+// Returns 0 if the address is not a staker
+func (sh *StakingHub) GetStakedBalance(address types.Address) *big.Int {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) {
+		return sh.StakingMap[address]
+	}
+
+	return big.NewInt(0)
+}
+
+// ComputeStakeAfterEvents goes over the pending events and returns the difference in value between staked and unstaked events
+// for the current address and block number
+func (sh *StakingHub) ComputeStakeAfterEvents(balance *big.Int, contextEvent PendingEvent) *big.Int {
+	stakedTally := balance
+
+	sh.EventQueueMutex.Lock()
+	defer sh.EventQueueMutex.Unlock()
+
+	// Go over events which have not yet been committed to find what will be
+	// the final staking tally for this address
+	for _, event := range sh.EventQueue {
+		// Check that we are working in the correct context
+		if event.BlockNumber != contextEvent.BlockNumber ||
+			event.Address.String() != contextEvent.Address.String() {
+			continue
+		}
+
+		if event.EventType == StakingEvent {
+			stakedTally.Add(stakedTally, event.Value)
+		} else if event.EventType == UnstakingEvent {
+			stakedTally = big.NewInt(0)
+		}
+	}
+
+	return stakedTally
+}
+
+// stakerMapping is a representation of a staked account balance
+type stakerMapping struct {
+	Address types.Address `json:"address"`
+	Stake   *big.Int      `json:"stake"`
+}
+
+// GetStakerMappings returns the staking addresses and their staking balances
+func (sh *StakingHub) getStakerMappings() []stakerMapping {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	mappings := make([]stakerMapping, len(sh.StakingMap))
+
+	var indx = 0
+	for address, stake := range sh.StakingMap {
+		mappings[indx] = stakerMapping{
+			Address: address,
+			Stake:   stake,
+		}
+		indx++
+	}
+
+	return mappings
+}
+
+// ClearEvents resets the event queue
+func (sh *StakingHub) ClearEvents() {
+	sh.EventQueueMutex.Lock()
+	defer sh.EventQueueMutex.Unlock()
+
+	sh.EventQueue = make([]PendingEvent, 0)
+}
+
+// ClearUnknownEvents purges the event queue of UnknownEvents
+func (sh *StakingHub) ClearUnknownEvents() {
+	sh.EventQueueMutex.Lock()
+	defer sh.EventQueueMutex.Unlock()
+
+	foundIndxs := make([]int, 0)
+	for indx, el := range sh.EventQueue {
+		if el.EventType == UnknownEvent {
+			foundIndxs = append(foundIndxs, indx)
+		}
+	}
+
+	if len(foundIndxs) > 0 {
+		prevSize := len(foundIndxs)
+
+		for _, foundIndx := range foundIndxs {
+			sh.EventQueue = append(sh.EventQueue[:foundIndx], sh.EventQueue[foundIndx+1:]...)
+		}
+
+		sh.log(fmt.Sprintf("Cleared [%d] unknown pending events", prevSize), logWarning)
+	}
```

Comment:
This might be more performant:
```suggestion
	knownEvents := make([]PendingEvent, 0)
	for _, el := range sh.EventQueue {
		if el.EventType != UnknownEvent {
			knownEvents = append(knownEvents, el)
		}
	}
	numUnknownEvents := len(sh.EventQueue) - len(knownEvents)
	if numUnknownEvents > 0 {
		sh.log(fmt.Sprintf("Cleared [%d] unknown pending events", numUnknownEvents), logWarning)
		sh.EventQueue = knownEvents
	}
```

Reply:
If you decide to keep this function, that is. Right now I believe it's not needed.

Reply:
Removed it as part of another comment 👍 
50f5bac